src/nvim/eval.c

```diff
@@ -5157,24 +5245,40 @@ void list_append_dict(list_T *list, dict_T *dict)
   ++dict->dv_refcount;
 }
 
-/*
- * Make a copy of "str" and append it as an item to list "l".
- * When "len" >= 0 use "str[len]".
- */
-void list_append_string(list_T *l, char_u *str, int len)
+/// Make a copy of "str" and append it as an item to list "l"
+///
+/// @param[out]  l    List to append to.
+/// @param[in]   str  String to append.
+/// @param[in]   len  Length of the appended string. May be negative, in this
+///                   case string is considered to be usual zero-terminated
+///                   string.
+void list_append_string(list_T *l, const char_u *str, int len)
+  FUNC_ATTR_NONNULL_ARG(1)
+{
+  if (str == NULL) {
+    list_append_allocated_string(l, NULL);
+  } else {
+    list_append_allocated_string(l, (len >= 0
+                                     ? xstrndup((char *) str, len)
```

Comment:
Use `xmemdupz` here? I always use this variant for performance reasons when I am pretty sure that there are either no embedded zero bytes or that they do not matter.